Remove deprecations

tiyash-basu-sociomantic opened the pull request.

tiyash-basu-sociomantic committed:
Remove usages of deprecated code

* Make imports being used transitively as `public`
Accessing modules transitively, i.e., accessing items in one module via an
item in another module, without publicly importing the first module into
the second, results in a deprecation.

* Use `ocean.meta.traits.Basic.isArrayType` instead of using
`ocean.core.Traits.isDynamicArrayType`,
`ocean.core.Traits.isStaticArrayType`, or `ocean.core.Traits.isArrayType`.

* Use `ocean.meta.traits.Basic.isReferenceType` instead of using
`ocean.core.Traits.isReferenceType`.

* Replace `ctfe_i2a` with `CTFE.toString`

* Replace `SizeofTuple` with `totalMemberSize`

* Replace `ContainsDynamicArray` with `containsDynamicArray`

* Replace `BaseTypeOfArrays` with `StripAllArrays`

* Replace `FieldType!(T, i)` with `typeof(T.tupleof[i])`

* Replace `FieldName!(i, T)` with `identifier!(T.tupleof[i])`

* Replace `ElementTypeOfArray` with `ElementTypeOf`

* Use `ocean.meta.types.Arrays.ElementTypeOf` instead of using
`ocean.core.Traits.ElementTypeOfArray`

* Use `ocean.meta.traits.Basic.isPrimitiveType`
instead of using `ocean.core.Traits.isPrimitiveType`

* Use `ocean.meta.traits.Indirections.hasIndirections`
instead of using `ocean.core.Traits.hasIndirections`.

* Use `ocean.meta.traits.Aggregates.hasMethod`
instead of using `ocean.core.Traits.hasMethod`.

* Use `ocean.meta.traits.Aggregates.hasMember`
instead of using `ocean.core.Traits.hasMember`.

* Directly copy objects as tuples using `dst.tupleof[] = src.tupleof[]`,
instead of using `copyClassFields`.

* Use `size_t` in foreach over array in `TimeHistogram`

* Use `ocean.meta.types.Templates.TemplateInstanceArgs`
instead of `ocean.core.Traits.TemplateInstanceArgs`.

tiyash-basu-sociomantic committed:
Remove deprecations since last major release

tiyash-basu-sociomantic committed:
Add more qualifiers to `toHash` methods

This addresses a warning for the specific instances of the methods, as
indicated in the diff.

The warnings addressed here look like:
`Warning: toHash() must be declared as extern (D) size_t toHash() const nothrow @safe, not const ulong()`

tiyash-basu-sociomantic committed:
Remove `old_handler` from `ConnectionHandler.RequestMap.RequestInfo`

This was unused post removal of `ConnectionHandler.RequestMap.add()` method.

tiyash-basu-sociomantic committed:
Remove unused interface `IRequestHandler`

from `swarm.neo.node.IRequestHandler` module,
and rename the module to `swarm.neo.node.IRequest`, because its remaining
content is the interface `IRequest`.

Also, added a release note here to indicate migration change.

tiyash-basu-sociomantic committed:
Move `in` contracts within function bodies

All `in` contracts were converted into `verify` calls in method bodies in
a previous PR. It seems some were missed. Converting these now.
(The reason for converting `in` contracts to `verify` calls is to make
important checks also occur in release builds.)

tiyash-basu-sociomantic committed:
Use cachalot version v5 in CI

tiyash-basu-sociomantic committed:
Add tests with dmd v2.085.* to CI

gavin-norman-sociomantic commented on src/swarm/neo/authentication/HmacAuthCode.d:
```diff
@@ -109,10 +109,7 @@ struct HmacAuthCode
-        mixin("shared static this ( ) { staticCtor(); }");
-    else
-        static this ( ) { staticCtor(); }
+    shared static this ( ) { staticCtor(); }
```
This could be refactored to include the code that's inside `staticCtor`. I don't think having that separate function is needed any more.

tiyash-basu-sociomantic replied:
I added a release note for this.

tiyash-basu-sociomantic replied:
Done!

gavin-norman-sociomantic commented on src/swarm/neo/node/Connection.d:
```diff
@@ -23,7 +23,7 @@ class Connection: ConnectionBase
     import swarm.neo.protocol.connect.NodeConnect;
     import swarm.neo.authentication.HmacDef: Key;
-    import swarm.neo.connection.YieldedRequestOnConns;
+    public import swarm.neo.connection.YieldedRequestOnConns;
```
Did you copy this decision (to use `public import`) from somewhere else (e.g. ocean)? It"s something we've traditionally avoided, but I may have missed that it's now being used in certain circumstances.

tiyash-basu-sociomantic replied:
This was not copied from any other project. I did not know that we are avoiding this.

Would you prefer to not use `public imports` here?

gavin-norman-sociomantic replied:
As far as I remember, in other library locations where we came across this, we added the imports in the locations where they're actually used.

tiyash-basu-sociomantic replied:
I have reverted back to `import` from `public import` in most cases, and imported modules explicitly, wherever needed.

In some places in `swarm`, I noticed that `public import`s are already being used, e.g., in `swarm.util.Verify`. I also found a few instances in `ocean`. So, in the [commit in this pr](https://github.com/sociomantic-tsunami/swarm/pull/393/commits/23d6a2b93abe9ac1907f4d17b9a53bbdafb7ee05), `public import`s are still introduced in some places in tests, where adding them keeps the design simplified.

gavin-norman-sociomantic commented on src/swarm/neo/client/ConnectionSet.d:
```diff
@@ -917,12 +918,8 @@ version (UnitTest)
 unittest
 {
     static void setAddress ( ref AddrPort address, ubyte[] address_bytes ... )
-    in
```
Are these changes related to the D2 transition?

tiyash-basu-sociomantic replied:
This was mentioned in the D2-migration docs [here](https://github.com/sociomantic/backend/blob/master/doc/d2-migration/stage3.md#invariants).

gavin-norman-sociomantic commented on src/swarm/client/registry/DisablableNodeSet.d:
```diff
@@ -83,16 +83,14 @@ public class DisablableNodeSet : NodeSet
     ***************************************************************************/
 
     public void disable ( NodeItem node )
-    out
```
Are these changes related to the D2 transition?

tiyash-basu-sociomantic replied:
This was mentioned in the D2-migration docs [here](https://github.com/sociomantic/backend/blob/master/doc/d2-migration/stage3.md#invariants).

gavin-norman-sociomantic replied:
The more detailed doc about assert/enforce/verify says something different: https://github.com/sociomantic-tsunami/sociomantic/blob/master/Code/D/assert-vs-enforce.md#adapting-older-code
>All checks in in contracts should be converted to verify statements in the method/function body.
>out contracts and invariants can remain as they are, using asserts. They will not be present in release builds.

The weird thing is that we already did a sweep of removing all `in` contracts from swarm, so I'm confused as to how you've discovered some!

tiyash-basu-sociomantic replied:
Interesting.. I have found some commits that remove `in` contracts, but the ones changed in this PR were not touched back then. Maybe they were missed for some reason?

Also, the `out` contracts have been put back now.

gavin-norman-sociomantic replied:
It _could_ be that "in contracts (containing asserts) should only be used on private methods/ functions which are called via a public API including verify/enforce checks. in contracts on non-private methods/functions should be avoided." applies to them? (https://github.com/sociomantic-tsunami/sociomantic/blob/master/Code/D/assert-vs-enforce.md#general-error-checking-principles)

gavin-norman-sociomantic commented on src/swarm/neo/node/IRequest.d:
```diff
@@ -0,0 +1,44 @@
+
+*******************************************************************************/
+
+module swarm.neo.node.IRequest;
```
This rename is a breaking change and should have a release note (for the sanity of library users).

tiyash-basu-sociomantic replied:
I added a release note for this.

gavin-norman-sociomantic reviewed (commented):
A few comments, but looks good!

tiyash-basu-sociomantic commented:
Marked #390 to be fixed in the commit titled `Remove D1 tests from CI`.

gavin-norman-sociomantic commented on src/swarm/neo/protocol/socket/MessageReceiver.d:
```diff
@@ -383,17 +383,15 @@ private class MessageReceiverBase
     }
     body
     {
+        verify(this !is null);
```
This is wrong. Doing `assert(this);` causes the invariant to be called. (Hence the comment.)

tiyash-basu-sociomantic replied:
Thanks for pointing out. Replaced the invariant call with the invariant body here.

gavin-norman-sociomantic replied:
Would be nice to not make this intermediary wrong change here.

gavin-norman-sociomantic commented on src/swarm/neo/protocol/socket/MessageReceiver.d:
```diff
@@ -515,16 +513,14 @@ class MessageReceiver: MessageReceiverBase
     }
     body
     {
+        verify(this !is null);
```
Ditto.

tiyash-basu-sociomantic replied:
Replaced the invariant call with the invariant body here.

gavin-norman-sociomantic replied:
Would be nice to not make this intermediary wrong change here.

tiyash-basu-sociomantic committed:
Remove deprecated usages of transitive imports

Unlike in D2, in D1, imports were public by default. In several places,
where imports were used transitively (i.e., C imports B, B imports A, and
C tries to use members from A without importing A),
result in deprecation warnings in D2.

This commit addresses those deprecations specifically. In unit tests and
system tests, imports have been made public, wherever it does not degrade
the existing design. Everywhere else, the necessary modules have been
explicitly imported.

matthias-wende-sociomantic commented:
Does all of this have to happen in one PR. Otherwise I'd suggest to have one PR for pure D2 migration and a succeeding cleanup PR.

gavin-norman-sociomantic commented:
That's what I was wondering. We could have a PR for the D2 stuff, then a PR for the deprecations, say.

gavin-norman-sociomantic commented on src/swarm/neo/client/RequestSet.d:
```diff
@@ -606,7 +606,13 @@ public final class RequestSet: IRequestSet
         out
         {
-            assert(this); // invariant
+            if (!this.id)
```
Wait, what's the reason for this?

tiyash-basu-sociomantic replied:
Thanks for pointing out. It seems like I missed it. I'll do another sweep.

tiyash-basu-sociomantic replied:
Ah, I see. I copied the contents of the invariant into the out block, so that checks specified here are not removed.

gavin-norman-sociomantic replied:
Why though? (This isn't what the assert/enforce/verify doc recommends doing. Unless I'm going mad.)

gavin-norman-sociomantic commented on src/swarm/client/request/params/IRequestParams.d:
```diff
@@ -538,7 +538,6 @@ public abstract class IRequestParams
         ***********************************************************************/
 
         static typeof(&this) serialized ( void[] data )
-        in
```
The commit message for this commit (" Move `in` contracts within function bodies ") is a bit inaccurate. Maybe this?
"
Move `in` contracts within function bodies
All `in` contracts were converted into `verify` calls in method bodies in a previous PR. It seems some were missed. Converting these now. (The reason for converting `in` contracts to `verify` calls is to make important checks also occur in release builds.)
"

tiyash-basu-sociomantic commented:
As discussed, this PR has been split into two. This PR takes care of some optimizations post-D2 migration, and removing deprecations.

ATM, it's blocked on https://github.com/sociomantic-tsunami/swarm/pull/395 to drop D1 support.

tiyash-basu-sociomantic commented:
Rebased on the `v6.x.x` branch.

gavin-norman-sociomantic reviewed (commented):
All commits which I've not commented on look good.

tiyash-basu-sociomantic commented:
As discussed, this commit refactoring `out` contracts has been removed, so that invariants are still executed in non-release builds.
